src/Type/Constant/ConstantArrayType.php

```diff
@@ -1212,26 +1212,32 @@ public function generalizeToArray(): Type
 	/**
 	 * @return self
 	 */
-	public function getKeysArray(): Type
+	public function getKeysArray(?Type $filterValueType = null): Type
 	{
-		$keyTypes = [];
-		$valueTypes = [];
-		$optionalKeys = [];
-		$autoIndex = 0;
+		$builder = ConstantArrayTypeBuilder::createEmpty();
 
 		foreach ($this->keyTypes as $i => $keyType) {
-			$keyTypes[] = new ConstantIntegerType($i);
-			$valueTypes[] = $keyType;
-			$autoIndex++;
+			$isOptionalKey = $this->isOptionalKey($i);
+			if ($filterValueType !== null) {
+				if ($filterValueType->isSuperTypeOf($this->valueTypes[$i])->no()) {
+					continue;
+				}
 
-			if (!$this->isOptionalKey($i)) {
-				continue;
+				if (!$filterValueType->equals($this->valueTypes[$i])) {
```

Comment:
This logic should be able to use isSuperTypeOf yes/no/maybe.

Reply:
this is an edge case I think, it does already above to skip values for `no`, but here (with `maybe|yes` remaining) I want to mark `maybe|yes` as optional EXCEPT they are actually the same. example: filtering with `bool` makes the key with value`true`, but filtering with `true` does not. does that make sense? I had a couple of such things already in the past, where I'm interested in basically `$foo->isSuperTypeOf($bar)->yes() && !$foo->equals($bar)`